rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,490 @@
+
+```ocaml
+val block_compare = lexicographic [
+  Comparable.lift Length.compare ~f:Block.block_length;
+  Comparable.lift Epoch.compare ~f:Block.epoch;
+  Comparable.left Epoch.Slot.compare ~f:Block.epoch_slot
+]
+```
+
+Additionally, we can order transactions by the minimum `block_compare` of the blocks that they appear in. Typically, a client would expect to receive a transaction at the chain that honest peers are building. This honest chain should have a small `block_compare` to other blocks. If the transactions do not appear in an block, they will not appear in the sorting.
+
+A drawback with sorting blocks and transaction based on the time that they are first seen is that if the client wants to store an arbitrary transaction and block, then they would not have the extra burden of tagging these objects with a timestamp for pagination. `block_compare` alleviates this issue.
+
+## SQL Database
+
+This architecture supports many SQL databases and it can even exclude a database entirely to make the architecture modular. When it is included, there are different use cases for using a light database, such as SQLite, and heavier database, such as Postgres. Postgres or even a distributed cloud database, such as AppSync, should be used when we have an archive node or a block explorer trying to make efficient writes to store everything that it hears from the network. These databases also support built-in subscriptions that would be convinient for clients to listen to.
+
+For the objects discussed in the previous section, we can embed them as tables in one global SQL database. Below are the schemas of the tables:
+
+```
+Table block {
+  state_hash string [pk]
+  creator string [not null]
+  staged_ledger_hash string [not null]
+  ledger_hash string [not null]
+  epoch int [not null]
+  slot int [not null]
+  ledger_proof_nonce int [not null]
+  status int [not null]
+  block_length int [not null]
+  block_time date [not null]
+  Indexes {
+    (block_length, epoch, slot) [name: "block_compare"]
+    block_time [name: "block_compare"]
+  }
+}
+
+  // Consensus_status is an ADT, but with its small state spaces, it can be coerced into an integer to make it easy to store it in a database. Here are the following states:
+  // - pending : The block is in the transition frontier and its value is its block confirmation number. Value[0, (k-1)]
+  // - confirmed : The block has reached finality by passing the root. Value=-1
+  // - failure: The block got removed from the transition frontier. Value=-2
+  // - unknown: After becoming online and bootstrapping, we may not know what will happen to the block. Value=-3
+
+Table user_commands {
+  id string [pk]
+  is_delegation bool [not null]
+  nonce int64 [not null]
+  sender string [not null]
+  receiver string [not null]
+  amount int64 [not null]
+  fee int64 [not null]
+  memo string [not null]
+  first_seen date
+  transaction_pool_membership bool [not null]
```

Comment:
We will handle the tracking of whenever a node bootstraps or restarts. We need to have the `transaction_pool_membership` because we need a way to track whether a transaction has a scheduled status or not. I don't there is an easy way to do that...